### ISSUE TITLE
QuickStart - Browse Data does not display content for certain URIs

### ISSUE THREAD

**Issue by bciano** [marklogic/marklogic-data-hub#557]
#### The issue

Short description of the problem:

Quick Start "Browse Data" does not display documents that include an "&" in the document URI

Application is not encoding the ampersand, thus submitting an invalid document URI in the request to the REST API

This works: http://localhost:8080/api/search/doc?database=FINAL&docUri=/XML_polis_3587737-GTA-00001.xml?dekcode=350**%26**objcode=856

What behavior are you expecting?
QS should properly encode the doc URI to be sent to the /doc ML REST api

#### Tech details

Which Operating System are you using?
mac osx

Which version of MarkLogic are you using?
9.0-2

Which version of the Data Hub Framework are you using?

Is this a QuickStart UI bug? If so, which browser are you using?

yes.  Chrome Version 64.0.3253.0

#### The devil is in the details...

If possible, rerun the command with -d for debugging output and attach the output:

05:04:31.367 [http-nio-8080-exec-1] ERROR o.a.c.c.C.[.[.[.[dispatcherServlet] - Servlet.service() for servlet [dispatcherServlet] in context with path [] threw exception [Request processing failed; nested exception is com.marklogic.client.ResourceNotFoundException: Local message: Could not read non-existent document. Server Message: RESTAPI-NODOCUMENT: (err:FOER0000) Resource or document does not exist:  category: content message: /XML_polis_3581860-GTA-00043.xml?dekcode=350] with root cause
com.marklogic.client.ResourceNotFoundException: Local message: Could not read non-existent document. Server Message: RESTAPI-NODOCUMENT: (err:FOER0000) Resource or document does not exist:  category: content message: /XML_polis_3581860-GTA-00043.xml?dekcode=350
	at com.marklogic.client.impl.OkHttpServices.getDocumentImpl(OkHttpServices.java:784)
	at com.marklogic.client.impl.OkHttpServices.getDocument(OkHttpServices.java:656)
	at com.marklogic.client.impl.DocumentManagerImpl.read(DocumentManagerImpl.java:389)
	at com.marklogic.client.impl.DocumentManagerImpl.read(DocumentManagerImpl.java:273)
	at com.marklogic.client.impl.DocumentManagerImpl.read(DocumentManagerImpl.java:239)
	at com.marklogic.client.impl.DocumentManagerImpl.readAs(DocumentManagerImpl.java:191)
	at com.marklogic.client.impl.GenericDocumentImpl.readAs(GenericDocumentImpl.java:23)
	at com.marklogic.client.impl.DocumentManagerImpl.readAs(DocumentManagerImpl.java:175)
	at com.marklogic.client.impl.GenericDocumentImpl.readAs(GenericDocumentImpl.java:23)
	at com.marklogic.quickstart.service.SearchService.getDoc(SearchService.java:112)
	at com.marklogic.quickstart.service.SearchService$$FastClassBySpringCGLIB$$ca9341e5.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:204)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:738)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157)
	at org.springframework.aop.support.DelegatingIntroductionInterceptor.doProceed(DelegatingIntroductionInterceptor.java:133)
	at org.springframework.aop.support.DelegatingIntroductionInterceptor.invoke(DelegatingIntroductionInterceptor.java:121)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:673)
	at com.marklogic.quickstart.service.SearchService$$EnhancerBySpringCGLIB$$d76d9573.getDoc(<generated>)
	at com.marklogic.quickstart.web.SearchController.getDoc(SearchController.java:57)
	at sun.reflect.GeneratedMethodAccessor179.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:205)
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:133)
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:97)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:827)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:738)
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:85)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:967)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:901)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:970)
	at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:861)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:687)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:846)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:317)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:127)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:91)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:114)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.session.SessionManagementFilter.doFilter(SessionManagementFilter.java:137)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.authentication.AnonymousAuthenticationFilter.doFilter(AnonymousAuthenticationFilter.java:111)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter.doFilter(SecurityContextHolderAwareRequestFilter.java:170)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.savedrequest.RequestCacheAwareFilter.doFilter(RequestCacheAwareFilter.java:63)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter.doFilter(AbstractAuthenticationProcessingFilter.java:200)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.authentication.logout.LogoutFilter.doFilter(LogoutFilter.java:116)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.header.HeaderWriterFilter.doFilterInternal(HeaderWriterFilter.java:64)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:105)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.context.request.async.WebAsyncManagerIntegrationFilter.doFilterInternal(WebAsyncManagerIntegrationFilter.java:56)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:214)
	at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:177)
	at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:346)
	at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:262)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:99)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.HttpPutFormContentFilter.doFilterInternal(HttpPutFormContentFilter.java:105)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.HiddenHttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:81)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:197)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:198)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:478)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:140)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:80)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:87)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:342)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:799)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:66)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:868)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1455)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:748)

┆Status: Done
┆Attachments: <a href="https:&#x2F;&#x2F;project.marklogic.com&#x2F;jira&#x2F;secure&#x2F;attachment&#x2F;35831&#x2F;Screen+Shot+2018-04-16+at+11.06.35+AM.png">Screen Shot 2018-04-16 at 11.06.35 AM.png</a>

**Comment by dmcassel** [comment]
➤ Scott Brooks commented:

After walking through this with Aries, there appears to be a more nuanced issue here relating to the contents of the database, which may point to an issue with the /doc endpoint as well. 

If you load only documents with ampersands in the URIs, then click on a document in the list, the endpoint throws an error. 

If you then load in more documents without ampersands in the URI, then click the SAME document that has the ampersands in the URI, the document contents render as it should.  

So, I need to dig deeper to see why the SAME REQUEST to the endpoint resolves fine in some cases and not in others.

Thanks [~ayuwono] for walking me through the reproduction of the issue.

**Comment by dmcassel** [comment]
➤ Aries Yuwono commented:

Additional information when walking through this with Scott

It seems that the string after the & is parsed as param

Request URL: http://localhost:4200/api/search/doc?database=STAGING&docUri=/board_games_extensions.csv-0-76&type=foo

Parsed query string params (from chrome network inspect):

  * database: STAGING
  * docUri: /board_games_extensions.csv-0-76
  * type: foo

When loading the documents through input flow, add "&type=foo" value to Output URI Suffix

**Comment by dmcassel** [comment]
➤ Scott Brooks commented:

Thanks [~ayuwono] .  The odd part is the request string is the same in the success and failure cases, so this leads me to believe that there might be an issue with the /api/search/doc endpoint, rather than with the encoding fix I put in.

**Comment by dmcassel** [comment]
➤ Scott Brooks commented:

Hey [~ayuwono], I'm looking at the latest code on the "develop" branch and am unable to reproduce what you're seeing.

See video here:

[https://drive.google.com/file/d/1Yw_qpmrfKfr71wxPRuMkSV2AfWbNrasF/view](https://drive.google.com/file/d/1Yw_qpmrfKfr71wxPRuMkSV2AfWbNrasF/view)

My parsed query string params (from chrome network inspect):

  * database: STAGING
  * docUri: /example.xml?blah=test

Could it be that you weren't running the latest code?  Give it another shot on the latest develop branch.  If you're still seeing failures, please document the process you used.  I noticed that you were running a flow to create the new documents in the Final database.  I simply inserted docs with and without URIs into each DB from Query Console... which should be equivalent.

**Comment by ayuwono** [comment]
Hi @clockworked247, this issue is about the "&" character, not "?" character

I tried it by inserting the doc on qconsole too (into STAGING db)

This works because the URI only contains "?" character:
```
'use strict';

declareUpdate();
xdmp.documentInsert('/doc2.json?type=foo', {foo:'new content here'})
```

This doesn't work because the URI contains "&" character:
```
'use strict';

declareUpdate();
xdmp.documentInsert('/doc2.json?doc=yes&type=foo', {foo:'new content here'})
```

**Comment by dmcassel** [comment]
➤ Scott Brooks commented:

bah, my fault.  You're right... looking again.

**Comment by dmcassel** [comment]
➤ Scott Brooks commented:

Sorry for the churn.  A new fix is in and the PR is in review:  [https://github.com/marklogic/marklogic-data-hub/pull/915](https://github.com/marklogic/marklogic-data-hub/pull/915)

**Comment by dmcassel** [comment]
➤ Scott Brooks commented:

Merged into the develop branch.  Passing back your way [~ayuwono]

**Comment by dmcassel** [comment]
➤ Aries Yuwono commented:

Verified and passed QA

**Comment by aebadirad** [comment]
Fixed with release of 3.0.0
